src/cppkin.h

```diff
@@ -14,87 +14,101 @@
-            cppkin::SpanContainer::Instance().SetSpan(span.release()); \
-        } \
-    }while(0)
-
-#define SERIALIZE_TRACING_HEADER(encodingType, outputData) \
-    do  \
-    {   \
-        if(cppkin::Sampler::ShouldSample()){ \
-            cppkin::Span& span = cppkin::SpanContainer::Instance().GetSpan(); \
-            cppkin::EncoderContext##encodingType context; \
-            cppkin::Encoder<EncodingType::encodingType>::Serialize(context, span.GetHeader()); \
-            outputData = std::move(context.ToString()); \
-        } \
-        else \
-            outputData = std::string(); \
-    }while(0)
-
-#define DESERIALIZE_TRACING_HEADER(decodingType, data, length, outputHeader) \
-    do  \
-    {   \
-        if(length > 0){ \
-            cppkin::EncoderContext##decodingType context(data, length); \
-            outputHeader = cppkin::Encoder<EncodingType::decodingType>::DeSerializeSpanHeader(context); \
-        } \
-    }while(0)
-
-#define SUBMIT_SPAN()  \
-    do  \
-    {   \
-        if(cppkin::Sampler::ShouldSample()){ \
-            std::unique_ptr<cppkin::Span> span = cppkin::SpanContainer::Instance().ReleaseSpan();   \
-            span->SetEndTime(); \
-            cppkin::TransportManager::Instance().PushSpan(std::move(span)); \
-        } \
-    }while(0)
-
-#define TRACE_EVENT(eventName) \
-    do  \
-    {   \
-        if(cppkin::Sampler::ShouldSample()){ \
-            cppkin::Span& span = cppkin::SpanContainer::Instance().GetSpan();   \
-            span.CreateSimpleAnnotation(eventName); \
-        } \
-    }while(0)
+class CppKin {
+public:
+    static void Init(const cppkin::CppkinParams& params) { cppkin::Init(params); }
+    static void CreateTrace(const char* operationName);
+    static void CreateSpan(const char* operationName);
+    static void JoinSpan(const cppkin::Span::SpanHeader& spanHeader);
+    static void TraceEvent(const char* eventName);
+    static void SubmitSpan();
+    static std::string SerializeTracingHeader();
+    static void SetTracingHeader(char* data, size_t length);
+
+    class SpanGuard {
+    public:
+        SpanGuard(const char* operationName) { CppKin::CreateSpan(operationName); }
+        ~SpanGuard() { CppKin::SubmitSpan(); }
+    };
+};
+
+inline void CppKin::CreateTrace(const char* operationName){
+    if(!cppkin::Sampler::AdvanceSampler())
+        return;
+    uint_fast64_t id = cppkin::Trace::Instance().GenerateTraceID();
+    std::unique_ptr<cppkin::Span> span = std::move(cppkin::Trace::Instance().CreateSpan(operationName, id));
+    cppkin::SpanContainer::Instance().PushSpan(span.release());
+}
+
+inline void CppKin::CreateSpan(const char* operationName){
+    cppkin::SpanContainer& spanContainer = cppkin::SpanContainer::Instance();
+    const cppkin::Span::SpanHeader* parentHeader = spanContainer.GetRootHeader();
+    if(!parentHeader) // No parent available. Span always has to have a parent. Skipping
```

Comment:
Question the current cycle and not the header, it doesn't seems right we question one time the cycle and the next the existence of the header while the logic of the previous already exists.